gustaf/spline/microstructure/generator.py

```diff
@@ -0,0 +1,553 @@
+                if self.tiling[i_pd] != 1:
+                    def_fun_para_space.insert_knots(
+                            i_pd,
+                            deformation_function_copy_.unique_knots[i_pd][1:-1]
+                    )
+            def_fun_para_space = def_fun_para_space.extract.beziers()
+
+        # Determine element resolution
+        element_resolutions = [
+                len(c) - 1 for c in deformation_function_copy_.unique_knots
+        ]
+
+        # Start actual composition
+        microstructure = []
+        if is_parametrized:
+            for i, (def_fun, def_fun_para) in enumerate(
+                    zip(def_fun_patches, def_fun_para_space)
+            ):
+                # Evaluate tile parameters
+                positions = def_fun_para.evaluate(
+                        self._microtile.evaluation_points
+                )
+                tile_parameters = self._parametrization_function(positions)
+
+                # Check if center or closing tile
+                if is_closed:
+                    # check index
+                    index = i
+                    for ipd in range(closing_face_dim):
+                        index -= index % element_resolutions[ipd]
+                        index /= element_resolutions[ipd]
+                    index = index % element_resolutions[closing_face_dim]
+                    if index == 0:
+                        # Closure at minimum id
+                        tile = self._microtile.closing_tile(
+                                parameters=tile_parameters,
+                                closure=closing_face + "_min",
+                                **kwargs
+                        )
+                    elif (index + 1) == element_resolutions[closing_face_dim]:
+                        # Closure at minimum id
+                        tile = self._microtile.closing_tile(
+                                parameters=tile_parameters,
+                                closure=closing_face + "_max",
+                                **kwargs
+                        )
+                    else:
+                        tile = self._microtile.create_tile(
+                                parameters=tile_parameters, **kwargs
+                        )
+                else:
+                    tile = self._microtile.create_tile(
+                            parameters=tile_parameters, **kwargs
+                    )
+
+                # Perform composition
+                for tile_patch in tile:
+                    microstructure.append(def_fun.compose(tile_patch))
+        # Not parametrized
+        else:
+            # Tile can be computed once (prevent to many evaluations)
+            tile = self._microtile.create_tile(*kwargs)
+            for def_fun in def_fun_patches:
+                for t in tile:
+                    microstructure.append(def_fun.compose(t))
+
+        return microstructure
+
+    def _sanity_check(self) -> bool:
```

Comment:
```suggestion
    def _sanity_check(self):
```
either full typing of none